proposals/0000-dependent-type-design.rst

```diff
@@ -962,6 +962,63 @@ will treat these two very differently:
 Even once we reach the Glorious Day, nothing forces us to unify ``t1 -> t2``
 with ``foreach (_ :: t1) -> t2``, and we may decide not to.
 
+Learnability of Haskell
+^^^^^^^^^^^^^^^^^^^^^^^
+
+A cross-cutting concern in the design of depdendent types in Haskell is whether
+they will make learning Haskell more difficult for newcomers to the language, or
+less usable for long-time Haskellers who prefer to avoid dependent types.
+
+We thus set forth this principle:
+
+**The Opt-In Principle (OIP):** Users who do not opt into dependent types will
+not be affected by them.
```

Comment:
I wanted to comment when this was first posted, but held off for a bit. Having waited and reviewed again, I still believe this is important to discuss, so here goes.

When I read over supportive posts in the PR, such as Tikhon's at https://github.com/ghc-proposals/ghc-proposals/pull/378#issuecomment-727149477 (though there are _many_ others where we can see this too), it's noted numerous times how DH is uncharted territory with significant unknowns. It's acknowledged as an experiment, and that things could go badly in various ways. That we'll need to iterate on this for quite a while before it'll be ready for general/production/widespread use. It's also noted that tutorials and docs the world over will need big updates. But then it's also noted that we couldn't possibly fork or work on the DH effort separately, to come back around into GHC in the future (though DH is a bit of a fork in its own right). It's noted that we can only "test" the soundness of DH by "doing it at scale" (eg on the haskell language/package community, whether they all want it or not).

I have a really difficult time accepting that DH will be "opt-in", or rather, that opting-out will be sufficient. As a newb, I would even argue that it's _not possible_ to opt-out.

For example.. let's say you want to work with envvars, you'll need to understand String/Text and IO. If you want to work with JSON, you'll be using aeson and will need to wrangle data types and records like a boss (you can get away with not knowing too much about type classes). IDK about you, but it's pretty common to start a new project, and almost immediately be faced with some JSON. Newbs will want to do this as a day 1 project. Let's say you want to interact with a few AWS managed services, also not a crazy request for a day 1 newb haskeller.. but now we're talking amazonka and lenses, which is a bit too much to take in (esp. the GHC errors).  Let's also be practical.. For common "real world" tasks completed in haskell (json, crypto, git, cloud APIs, TLS, etc), there are only 1 - 3 "options" for libraries. As a user of those libraries, I have to use/learn/understand/master whatever language features the library enables and/or demands for use.

So, sure, we can hide DH behind some language pragma, but the opt in/out is much more complicated than that, particularly in a community of Haskell's size (larger communities could more easily support multiple sub-ecosystems for how to do things, haskell hasn't fared well even with the stack/cabal split, so I don't see how this would be better).

In addition to that, GHC seems to have a significant gap in testing language features. With all the pragmas that are available, there needs to be more stringent and structured testing of the results from GHC with the multitude of combinations of features enabled (GHC test suite would need a matrix build against all those options). So until that changes, the subtle bugs and performance regressions will continue to elude us as well.

While this opt-in principal might be a nice goal or desire, it isn't grounded in a practical reality working with haskell today, it doesn't fit the reality of learning and using haskell in practice.

Reply:
I think including the greater ecosystem in the definition of "opt-in" sets an impossible bar. The principal already rightly focuses on what GHC can control: The compiler itself and `base`.

If you follow this greater opt-in principal to its natural conclusion, DTs are hamstrung from the get-go. The root cause of the "bad" situation you describe is exactly DTs being useful & desired by many Haskellers in the ecosystem! If DTs are useful, people will write libraries in them. But if people write libraries in them, they might end up being best-in-class and the "newb" may have to use them since the next best alternative isn't as good?

-----

Also, the fear mentioned here also doesn't even pan out in reality.

Say we have DTs and - like you say - there is now library-for-X that uses DTs and there isn't another library-for-X that doesn't use DTs in the ecosystem. So we have now arrived at the supposed violation of the opt-in policy. Okay, let's try not to violate it as an omniscient, omnipotent being. We have two options:

1. Write the non-DT library-for-X ourselves (or otherwise fund it.) Problem solved.
2. Rewrite history and never add DTs to GHC. Now instead of the only library-for-X "forcing" me to use DTs, there is no library-for-X at all! So we still have to do (1)

Reply:
This is a fair worry. There is a chance that DT will become so useful so quickly that libraries will flock to it and then users will have no choice (in practice) but to embrace it. Such an ecosystem would make the OIP rather moot.

On the other hand, we can do a better job than we have historically done around, say, error messages. For example, it would be easy to make sure that the first few releases of DT never have an error message which suggests to enable the extension -- or perhaps only to do so with warnings about the experimental nature of the extension. This is the area in which the OIP is easiest to hew to.

In the end, though, @ketzacoatl's argument here is a general argument against progress. That probably reads considerably more harshly that it's meant -- I definitely see the potential problems raised and share some of the concerns. However, if Haskell is to remain a language that welcomes experiments around its margins (as it has been since its inception), we need to remain hopeful that new features are indeed opt-in and that the community will uphold this principle.

I recognize that I'm saying "I promise we'll keep our experiments to ourselves" and @ketzacoatl is saying "I'm worried you won't be able to". Neither of us can substantiate our arguments, and only time will tell who is right (probably: both of us, to varying degrees). So it comes down to a question of the community's stance toward new ideas: do we want to keep experimenting? or do we declare ourselves essentially done and happy?

Reply:
> "I'm worried you won't be able to"

Less worry and more like: That's my present experience. I'm simply sharing what it's like for some of us. You can say it isn't or won't happen, but that's not the point.

Learning to be effective and proficient with haskell is an interesting experience. It takes time and dedication, and for whatever reason, it seems to be pretty universal that as we progress, we forget the previous mental states we had that carried us through earlier periods of learning. We forget why particular errors, mistakes, or misunderstandings were difficult for us at a particular time. We forget even, how to pass on our insights and understandings to colleagues with a significant gap in experience (relative to our present state). While there may be many, many skilled Haskellers, there are far fewer mentors in the haskell community that have a high level of experience, _and_ can teach any level haskeller. Most of us are the most efficient teaching a haskeller that is close to us in ability, and we are terrible at teaching haskellers further off from our abilities. It's like, as we grow up, we forget what it was like to be a kid.


>  Neither of us can substantiate our arguments

I disagree, quite a few issues raised are based on present/past experience with GHC and the Haskell ecosystem. Matt Parson's posts were able to provide more precise details than I, but I add my voice to his because I have experienced it first hand as he described.

> So it comes down to a question of the community's stance toward new ideas: do we want to keep experimenting? or do we declare ourselves essentially done and happy?

This isn't an attempt to shoot down your idea, it's feedback to help you do your work better. Just like applicative-do and <- make my life better (I can't even fathom how much), if DH makes my life writing software with Haskell better, of course I'm going to want it. But it's also ok to object to a really bumpy and unenjoyable ride while we get there. Or to point, for a moment, at how we can't really opt-out as much as is implied here (otherwise, we wouldn't consider this such a significant change, the whole point of the proposal is that it has very far-reaching consequences).

> I think including the greater ecosystem in the definition of "opt-in" sets an impossible bar. The principle already rightly focuses on what GHC can control: The compiler itself and base.

Part of my point is that, at least with what I've observed and experienced with Haskell, you cannot effectively separate those two. This "impossible bar" is exactly right, which is why this opt-in/out is unrealistic.

> If you follow this greater opt-in principle to its natural conclusion, DTs are hamstrung from the get-go. The root cause of the "bad" situation you describe is exactly DTs being useful & desired by many Haskellers in the ecosystem! If DTs are useful, people will write libraries in them. But if people write libraries in them, they might end up being best-in-class and the "newb" may have to use them since the next best alternative isn't as good?

The problem is that not only is this a very significant change to the language, but DH is unrefined and there are still so many unknowns. In an ideal world, the wizards would figure it out independently of the muggles continuing to use haskell in its current form, until DH is refined and ready for prime time. The DH wizards reply back they don't know how without "running the experiment on a library ecosystem at scale" (maybe you can understand that's a lot of faith/trust you are asking for). That, we could have DH without significant breakage or regression, I think that's a worthy thing to target. I'd like to point out that we are not as insulated from that work as you might like. At least, feel free to use this as feedback that helps you refine your course of action. I'd love it if you'd prove me wrong, with a smooth transition and absolutely fantastic user-experience.

Reply:
> Not really worry, more like: That's my present experience. I'm simply sharing what it's like for some of us. You can say it isn't or won't happen, but that's not the point.

To be clear, I'm not claiming (and I hope I have not previously claimed) that this situation doesn't exist today. I believe that this your experience, and I believe that GHC (and beyond) has fault to bear for creating this experience.

My claim is that the implementation of DT will make a hard try to break this trend -- this is the claim that I cannot substantiate and you cannot refute. Why do I think DT can be different? Precisely because we're having this conversation.

From its inception until [January 3, 2017](https://github.com/ghc-proposals/ghc-proposals/commit/9432ba75015cb9ab48e5aa7c8f6d7b626d1c07a6#diff-748800454a3e59d50289ba89776a95a67a66cefebaefa81f16009c58a4b28707), GHC was not controlled by any proposals process. When a developer had a fun idea (or a researcher developed an idea and then begged a developer to implement it), they circulated it among other devs, and eventually the idea made its way into GHC. The people discussing the idea were all programming-language-theory and Haskell experts. And so new features were designed, implicitly, to work well for programming-language-theory and Haskell experts. No novices were in the conversation to be confused by error messages or to talk about learnability (though developers often did take learnability into account). (I'm generalizing here and there are exceptions.)

Now, however, we have the opportunity to have these conversations, and @ketzacoatl is by far not the only one to be concerned by the impact of DT on non-DT programmers. This concern has been taken to heart and is enshrined in the OIP. Does this statement of the OIP mean that all will be well? Not at all, of course: care will have to be taken in the design and implementation of specific DT features. But I -- and others -- have heard loud and clear that, at least at first, continuity of experience should be prioritized over discoverability. That is, we should prefer to preserve old error messages instead of writing new ones that suggest enabling `-XDependentTypes` (for example). I think this is something we can achieve, and we have a chance to do this better than it has been done before precisely because we are more aware of the potential pitfalls.

>  it's feedback to help you do your work better

Indeed! I fully agree here. That is the spirit in which I take this.

> it's also ok to object to a really bumpy and unenjoyable ride while we get there

While it's always OK to object, I don't think we can (or should) try to avoid bumpy rides. Well, I don't quite mean that: we should try to avoid bumpy rides, but we should also expect to have bumps anyway. New features are, well, new. And unless we have an alternative Haskell ecosystem where a few thousand of us can experiment and grow libraries to use a new feature, all while maintaining the primary ecosystem, there will be hard lessons learned. That's simply not possible, and so we accept that some libraries may be early adopters of DT and might make mistakes. My hope is that libraries continue to support non-DT interfaces, but that's well beyond my (or any one person's/group's) control.

> maybe you can understand that's a lot of faith/trust you are asking for

Yes and no. Yes, it's a lot a faith that I'm asking for. But I also think it's a reasonable ask -- because you're getting something in return. You get to work in a language supported by bleeding-edge theory and its powerful features. In some sense, signing onto using Haskell means signing onto the occasional bump as new features get added. Unlike some other languages, Haskell is unafraid to innovate in this way, both to its credit and, sometimes, to its detriment. I think the value proposition here is good, but others may differ -- and that's fine, too.

> That, we could have DH without significant breakage or regression, I think that's a worthy thing to target.

Yes, absolutely! That's what we *are* targeting, as written in the OIP. I'm confident we can do this correctly within GHC and `base`. What happens in the library ecosystem is up to library authors.

Reply:
> But I also think it's a reasonable ask

I disagree. It may seem reasonable, but it's also pretty highly unreasonable from other perspectives, and part of the problem here is rooted in the fact that why this is unreasonable is a difficult perspective for the vocal proponents to embody and fully understand (I really appreciate that you care enough to do your best with that).

Here are a few thoughts on what I find unreasonable:

There are far fewer people contributing to this discussion / idea than those that will be affected (maybe it's a factor of 100 to 1000x, IDK). As you have described, the overwhelming majority of those contributing to GHC directly or in this discussion here are experienced/interested in type theory and the academics involved - you enjoy it. While that's great, it also skews the perspective we bring to discussion.

Similarly, experienced Haskellers who disagree have been _less vocal_ in objecting, than what they feel internally or have expressed privately. I would guess this is because dissent has been equated to "impeding progress", "contributing to brain drain", and "against change", and there is a cost to vocalizing disagreement with the majority.

Proponents are rather highly confident it'll be "for the better", and yet it's also more than a significant change to Haskell - DH is bordering on a new language, or even a fork of Haskell (I would guess proponents will flame me for saying this, but let's be realistic, this is a big change with far-reaching consequences, and is more similar to the Python 2/3 change than the typical new "language feature").

As you and others have described several times, this is unproven work, and which "cannot be proven, except at scale". That's a really difficult position to put people like me in. This is a tall ask. You've chosen Haskell for specific reasons, one being the experimental nature of haskell, and another being the _size_ of the ecosystem. The proposal is effectively asking if it's ok to run this _experiement_ on the entirety of the Haskell ecosystem and community. I object to that fundamental premise.


>  because you're getting something in return.

Part of the problem is the result you may think I'm getting and the result I get, are not necessarily the same. You look at the end result, I'm looking at the whole path to get there (which you are arguably too close to see without distortion, and I'm too far from to have all the pieces). You are an implementer and I am a user, and you believe you are doing this "for the good" and "to improve dev's experiences creating more robust software". What we believe we're getting is going to be fundamentally different. There is also the assumption that this gift is universally desired/accepted/wanted.


>  You get to work in a language supported by bleeding-edge theory and its powerful features.

This is part of the problem. I use Haskell for its robustness and correctness. STM is refined and robust, it just works, and probably took time to get there. STM is a gift. It's amazing but I didn't use STM while it was being refined. And it didn't interfere with my work during that time either.

I trust you when you say DH will _eventually_ be so awesome we'll all use it, and we won't even realise what we're using (just like STM is now). That DH would be a gift I'd want to experience. STM just works, imposes a _very small_ barrier to entry, and just doesn't get in my way. I'd be more than happy to use DH at that time, as I use STM now.

Overall, Haskell can already provide a better dev experience (by far), than all the other languages/tools in wide-scale, production use (Rust is closely catching up).


> In some sense, signing onto using Haskell means signing onto the occasional bump as new features get added. 

We've made it incredibly difficult to "fix" base, we have warts we say cannot change, "because", yet GHC marches on to its own beat on breaking changes irregardless of that. Putting aside that discrepancy, I  don't think it's reasonable to compare DH to those occasional bumps, these are not the same thing.


> Unlike some other languages, Haskell is unafraid to innovate in this way, both to its credit and, sometimes, to its detriment. 

I'm not against progress or change, I'm objecting to _how_ those changes are being made here.

Why can't DH figure itself out, become awesome, and let us migrate to it when we're ready and want to, and without taking us all on the rollercoaster?

Reply:
I don't really see the problem here. There are basically two things:
- stuff that is mature and useful (e.g. STM). They won't change because of DH, and still be as useful as today
- new stuff that becomes possible (or ergonomic enough) now that might create new libraries, then again we habe to possibilities:
    -DH is not perfect yet or the library auther has not found the ideal way to use DH. The library then might be clumsy to use, but then just dont do that
    - DH is really good and there is a new library that becomes best-in-class like it happened with servant. Then again we need to see two points:
        - There still will be alternatives, e.g. today many people use spock or yesod because servant is complex
        - If it is best-in-class it might be becuse it is easier to use than the alternatives, e.g. servant cuts down on the boilerplate and makes e.g. swagger a lot easier

So yeah, the only "danger" is if DH is good *and* can be used effectively *and* your company chooses to use DH. At that point I think DH has proven itself useful

Reply:
First off, well-argued @ketzacoatl. You are right that commenters here are far from representative, and that speaking up against DH in this thread is risky. I don't really know how to combat that, other than trying my best (it's not hard) to be welcoming to alternative viewpoints. Nevertheless, you are right to highlight this state of facts.

> Why can't DH figure itself out, become awesome, and let us migrate to it when we're ready and want to, and without taking us all on the rollercoaster?

How do you propose to do this? I have no idea. But without an answer to this question, your argument is, essentially, "let's not make any big changes to Haskell ever again" (for some definition of "big"). I'm really not trying to invalidate your argument (and certainly not your experience) -- just exploring what accepting your argument would imply.

For what it's worth, I think @jvanbruegge has a good breakdown of this, provided that GHC can keep to itself (which I'm confident it can).

Reply:
(I wasn't going to comment further than my already voluminous postings around DH, but some advice for @ketzacoatl's concerns -- which I share.)

> it's noted numerous times how DH is uncharted territory with significant unknowns.

My take is a) it's going to be a long time before DH comes together enough to be worth considering for intermediate-level Haskellers; b) during that time there will be no other development of any other features [**].

So my advice is to stick with whatever version of GHC and libraries are working for you now. They won't actually go backwards. Come back to GHC in (say) 5 year's time. Hope there's no brokenness in what you're using now that forces you to upgrade to a DH-tainted release. If you're in academe teaching Haskell, GHC vintage 2015 is a lot closer to all the intro texts, so will be more grokkable for newbies.

[**] It would be nice to see progress on `FunctionalDependencies`, `OverlappingInstances`, `UndecidableInstances`, more ergonomic `PatternSynonyms`, better integrating something like Liquid Haskell, ... -- there's heaps of ideas if you look back over proposals; there's even proposals that have been accepted but then got stuck/probably will never be implemented. I feel particularly that better `FunctionalDependencies` might be a less intellectually challenging competitor with comparable expressivity to some features of DH. But we'll never know. I guess nobody's going to get academic kudos/publishable papers from polishing up what is a too-workmanlike feature.

Reply:
Yep, exactly, and I believe that's an unreasonable request to make of the community at large.

Reply:
@jvanbruegge, the problem is more about the transition from here to "DH is stable". I would guess STM's development didn't disrupt anything major while it was refined. DH's impact is far more reaching, and will be harder to "contain" during the transition. This is why I push back and say: "why hasn't this proposal found a viable alternative route we can consider instead of tearing up the road we're all driving on"?

Reply:
@goldfirere, I am so glad that is the topic you have called out for review, and I really appreciate having this discussion. Having spent some time on it, I would guess that any possibilities I propose will be pushed back on as "too much work", otherwise the proponents here would have considered it further and probably even said that you would do it here. So that is a problem that inhibits our ability to find/evaluate reasonable possibilities.

While @AntC2 touched on it above, I would also like to focus a little more attention on it, because I believe that helps set better boundaries for success.. Seeing DH to "production quality" is a deliverable that is on the 5 - 10 year timeframe. With that in mind, we should plan for this type of transition period, and we should be practical in our analysis of what that means for the Haskell ecosystem.

That said, I'll share 2 insights I have on this topic:
* As an operator of things, I have to make changes with minimal disruption. We use many tricks to keep dev and product teams moving while we change out the tires or engine (all while the machine is still barreling down the highway). When we do it well, they barely know the change has happened, and thank us once they've caught up to understand the new reality they get to make use of. When our job is done well, you would even think we might not be necessary, the bumps come when we're not doing our work (minor blips are not bumps). By "bumps" I mean a network team installing a 'security device' on a Friday afternoon with no notice, and with insignificant prior testing, and which breaks a whole bunch of workflows actively in use ("in production", so to speak). I would like to challenge us to have a more graceful story and result for this major change to the Haskell language and community at large (reduce the bumps you expect to see down to blips we can all reasonably live with).
* While thinking over practical methods and ways that could apply here, I realized the following: DH needs space to more freely experiment, iterate, break, get stuck, redo itself, figuring itself out, and to be refined. **Attempting to define, iterate, and redefine DH in the current space that GHC provides seems too constrained for what DH needs. DH should have the freedom to iterate faster than GHC can allow. In fact, I think that is the faster and easier way to get a better DH product.** The challenge becomes setting up the right environment for that experimentation and development, with proper bridges to the GHC ecosystem and Haskell community at large.

Reply:
@ketzacoatl I agree with rae that your points are very well put. And you make a very good point about how huge changes with potentially negative effects for the large user base of GHC have to be taken with great care. That means acting on this proposal, if it gets accepted, will require paying great attention to these concerns.
But there is the claim here that developing DH in ghc is "taking us all on the rollercoaster". And while this is certainly an at least theoretically possible problem (especially if not considered from the beginning), I don‘t see very much evidence that this will happen:

* It does not sound very hard to keep DH safely behind one or multiple feature flags.
* It is perfectly possible to only show recommendations to enable some flag if the user has somehow signaled that they are in principle okay with such features.
* And I really can‘t imagine that important libraries in the ecosystem will jump on to the DH train before it has proven it’s worth. (And I actually expect that the more adventurous libraries will probably become easier to use in the long run.)

Again, I very much agree with your concerns. I am just not sold that the compromise is to not do this. I think "Why can't DH figure itself out, become awesome, and let us migrate to it when we're ready and want to, and without taking us all on the rollercoaster?" is exactly the right question. And I would be very interested in hearing why we can‘t have them with this proposal. Why do you expect the implementers of this proposal to fail at the reasonably easy steps (compared to the larger endeavor) outlined above against their declared intentions? Or where do you think these steps won‘t suffice?
The nice thing here is, btw., that every concrete argument how DH could fail to be self contained will either help in preventing this proposal or will improve the result.

Reply:
> DH should have the freedom to iterate faster than GHC can allow. In fact, I think that is the faster and easier way to get a better DH product. The challenge becomes setting up the right environment for that experimentation and development, with proper bridges to the GHC ecosystem and Haskell community at large.

I'm pretty sure a feature branch of GHC is the best solution to this. Interested super-early adopters can use the branch themselves without much trouble, and it of course naturally bridges to GHC because it's GHC.

Of course very few would use a branch to build production software, but that's also the stated purpose of the branch. It's a place to iterate fast, explore, and learn. Once the branch is merged, it'll be at least production-ready for the bravest at least.

Reply:
Also, I want to add something about this opaque term: "production quality"

The bar for "production quality" for a type-system feature is extremely low compared to - say - a database library or tweak to the RTS or a new compiler optimization.

For instance, `singletons` (feat of Haskelling it is) is below the bar of where some would consider DH "production quality." But deploying a program `singletons` to production isn't especially scary. I've done it - works great.

I'd say the same goes for the new `-XLinearTypes`. It is extremely new & experimental still. It has some ergonomic issues, and I'm guessing there will be GHC updates and proposals to improve it going forward. But I would not lose any sleep deploying code with modules enabling `LinearTypes` to a server serving millions of users of traffic.

And LTs seems to follow this OIP! And they follow it for the exact reasons we've been discussing: A need to experiment and not be bound to a brutally stable public API. Now, I could go spend a few months building libraries with no peer using LTs, but that wouldn't mean the extension now fails this OIP test.

But maybe the OIP definition itself should make 100% clear that third-party libraries and ecosystem are out of the OIP's scope. That's fair.